### ISSUE TITLE
Please consider a global-sharedscripts option for a single-shot processing script.

### ISSUE THREAD

**Issue by james-fitzrandolph** [logrotate/logrotate#432]
Please consider an option for globally-sharedscripts which would run the the script only after ALL specified logs have been run.  Given the following (simplistic) config:

```
sharedscripts [[ does not really do anything ]]
/var/log/daemon {
        size = 80000
        postrotate
                /bin/killall -HUP syslogd
        endscript
}
/var/log/debug {
        size = 25000
        postrotate
                /bin/killall -HUP syslogd
        endscript
}
/var/log/maillog {
        size = 50000
        postrotate
                /bin/killall -HUP syslogd
        endscript
}
```

If all three log files are rotated, syslogd will receive 3 HUP signals.  It would be nice if it could receive only 1 (true, the overhead in stopping/starting it is fairly minimal, but it is still there).

I am processing several syslog scripts in real time by utilizing inotify to let me know when new log messages have been added.  I have noticed that when logrotate runs, I quite frequently have multiple IN_CLOSE_WRITE notifications showing up for a single file.  I have had to add logic to determine that underlying inode has actually changed before processing what I originally thought should be a single close...

**Comment by kdudka** [comment]
Something like this has already been proposed as #424.

**Comment by james-fitzrandolph** [comment]
Not quite.  I want something that only gets called when logs have been rotated, but only called once no matter how many logs get rotated.  Mine will get called when logrotate has finished all processing related to log files, then call for additional processing just before it shuts down.

**Comment by james-fitzrandolph** [comment]
More information:

If I start with the following logrotate input file:
```
sharedscripts
/tmp/lr/loga /tmp/lr/logb {
  size = 8000
  postrotate
    /tmp/lr/killhupsyslog
  endscript
}
```

and the following script:
```
#! /bin/bash
date >> /tmp/lr/killhuplr.log
```
And a couple of log files that need rotating.  When I run logrotate I get:

```
caorach{jim}: logrotate -v -f input -s state
reading config file input
Reading state from file: state
Allocating hash table for state file, size 64 entries
Creating new state

Handling 1 logs

rotating pattern: /tmp/lr/loga /tmp/lr/logb  forced from command line (no old logs will be kept)
empty log files are rotated, old logs are removed
considering log /tmp/lr/loga
  Now: 2022-01-04 08:36
  Last rotated at 2022-01-04 08:35
  log needs rotating
considering log /tmp/lr/logb
Creating new state
  Now: 2022-01-04 08:36
  Last rotated at 2022-01-04 08:00
  log needs rotating
rotating log /tmp/lr/loga, log->rotateCount is 0
dateext suffix '-20220104'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/lr/loga.1 to /tmp/lr/loga.2 (rotatecount 1, logstart 1, i 1),
renaming /tmp/lr/loga.0 to /tmp/lr/loga.1 (rotatecount 1, logstart 1, i 0),
old log /tmp/lr/loga.0 does not exist
rotating log /tmp/lr/logb, log->rotateCount is 0
dateext suffix '-20220104'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/lr/logb.1 to /tmp/lr/logb.2 (rotatecount 1, logstart 1, i 1),
renaming /tmp/lr/logb.0 to /tmp/lr/logb.1 (rotatecount 1, logstart 1, i 0),
old log /tmp/lr/logb.0 does not exist
renaming /tmp/lr/loga to /tmp/lr/loga.1
disposeName will be /tmp/lr/loga.1
renaming /tmp/lr/logb to /tmp/lr/logb.1
disposeName will be /tmp/lr/logb.1
running postrotate script                          <---
removing old log /tmp/lr/loga.1
removing old log /tmp/lr/logb.1
```
This is great.  I get one run of the postrotate script.  I verify that by:
```
caorach{jim}: cat killhuplr.log
Tue Jan  4 08:36:53 MST 2022
```

This is great as long as the two files (loga and logb) have the same rotation requirements.  However, if loga needs to be rotated at a different max size, I need to rewrite the input as:
```
sharedscripts
/tmp/lr/loga {
  size = 8000
  postrotate
    /tmp/lr/killhupsyslog
  endscript
}
/tmp/lr/logb {
  size = 3000
  postrotate
    /tmp/lr/killhupsyslog
  endscript
}
```
When I run logrotate now, I get:
```
caorach{jim}: logrotate -v -f input -s state
reading config file input
Reading state from file: state
Allocating hash table for state file, size 64 entries
Creating new state
Creating new state

Handling 2 logs

rotating pattern: /tmp/lr/loga  forced from command line (no old logs will be kept)
empty log files are rotated, old logs are removed
considering log /tmp/lr/loga
  Now: 2022-01-04 08:58
  Last rotated at 2022-01-04 08:47
  log needs rotating
rotating log /tmp/lr/loga, log->rotateCount is 0
dateext suffix '-20220104'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/lr/loga.1 to /tmp/lr/loga.2 (rotatecount 1, logstart 1, i 1),
renaming /tmp/lr/loga.0 to /tmp/lr/loga.1 (rotatecount 1, logstart 1, i 0),
old log /tmp/lr/loga.0 does not exist
renaming /tmp/lr/loga to /tmp/lr/loga.1
disposeName will be /tmp/lr/loga.1
running postrotate script                           <---
removing old log /tmp/lr/loga.1

rotating pattern: /tmp/lr/logb  forced from command line (no old logs will be kept)
empty log files are rotated, old logs are removed
considering log /tmp/lr/logb
  Now: 2022-01-04 08:58
  Last rotated at 2022-01-04 08:47
  log needs rotating
rotating log /tmp/lr/logb, log->rotateCount is 0
dateext suffix '-20220104'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/lr/logb.1 to /tmp/lr/logb.2 (rotatecount 1, logstart 1, i 1),
renaming /tmp/lr/logb.0 to /tmp/lr/logb.1 (rotatecount 1, logstart 1, i 0),
old log /tmp/lr/logb.0 does not exist
renaming /tmp/lr/logb to /tmp/lr/logb.1
disposeName will be /tmp/lr/logb.1
running postrotate script      	       	       	    <---
removing old log /tmp/lr/logb.1
```
My post rotation script has been run twice.  Verify by:
```
caorach{jim}: cat killhuplr.log
Tue Jan  4 08:36:53 MST 2022
Tue Jan  4 08:47:09 MST 2022   <--- new
Tue Jan  4 08:47:09 MST 2022   <--- new
```

Had my post rotation script been an actual 'killall -HUP syslogd', it could potentially have sent 14 (on my system) different signals (all HUP) in basically zero time to the system logger.  I really only need one of them.

What I would like is something like this input file:
```
sharedscripts
/tmp/lr/loga {
  size = 8000
  oneshot
    /tmp/lr/killhupsyslog
  endscript
}
/tmp/lr/logb {
  size = 3000
  oneshot
    /tmp/lr/killhupsyslog
  endscript
}
```
Which generates output like:
```
caorach{jim}: logrotate -v -f input -s state
reading config file input
Reading state from file: state
Allocating hash table for state file, size 64 entries
Creating new state
Creating new state

Handling 2 logs

rotating pattern: /tmp/lr/loga  forced from command line (no old logs will be kept)
empty log files are rotated, old logs are removed
considering log /tmp/lr/loga
  Now: 2022-01-04 08:58
  Last rotated at 2022-01-04 08:47
  log needs rotating
rotating log /tmp/lr/loga, log->rotateCount is 0
dateext suffix '-20220104'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/lr/loga.1 to /tmp/lr/loga.2 (rotatecount 1, logstart 1, i 1),
renaming /tmp/lr/loga.0 to /tmp/lr/loga.1 (rotatecount 1, logstart 1, i 0),
old log /tmp/lr/loga.0 does not exist
renaming /tmp/lr/loga to /tmp/lr/loga.1
disposeName will be /tmp/lr/loga.1
defer running postrotate script                        	 <---
removing old log /tmp/lr/loga.1

rotating pattern: /tmp/lr/logb  forced from command line (no old logs will be kept)
empty log files are rotated, old logs are removed
considering log /tmp/lr/logb
  Now: 2022-01-04 08:58
  Last rotated at 2022-01-04 08:47
  log needs rotating
rotating log /tmp/lr/logb, log->rotateCount is 0
dateext suffix '-20220104'
glob pattern '-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
renaming /tmp/lr/logb.1 to /tmp/lr/logb.2 (rotatecount 1, logstart 1, i 1),
renaming /tmp/lr/logb.0 to /tmp/lr/logb.1 (rotatecount 1, logstart 1, i 0),
old log /tmp/lr/logb.0 does not exist
renaming /tmp/lr/logb to /tmp/lr/logb.1
disposeName will be /tmp/lr/logb.1
defer running postrotate script    	       	              <---
removing old log /tmp/lr/logb.1

running	deferred oneshot script
```
ie, what I am looking for is a 'sharedscript' option that is global across all rotation recipes.

**Comment by kdudka** [comment]
How would logrotate know which `oneshot` scripts are in the same group?  `/etc/logrotate.conf` usually contains `include /etc/logrotate.d`, which is a directory where independently maintained software projects put there own configuration files for logrotate.  These configuration files need to work more or less independently of each other, yet the administrator should be able to set system-wide logrotate options where appropriate.  I do not think that the proposed extension would work well in this ecosystem.
